### DISCUSSION TITLE
Why vllm-backend prepends output with the prompt?

### DISCUSSION THREAD

**Discussion by mkhludnev** [triton-inference-server/server#6867] [Q&A] [answered]
Here's [the concatenation](https://github.com/triton-inference-server/vllm_backend/pull/31/files)
Here's how output looks like for [REST and gRPC](https://github.com/langchain-ai/langchain/issues/12474#issuecomment-1925914483) 
[Bug](https://github.com/triton-inference-server/server/issues/6866) on server

**Comment by oandreeva-nv** [comment]
I agree that we most likely need to remove `prompt` concatenation and just send the generated response. Alternatively, we can introduce `keep prompt` flag in `config` and default it to `false`. CC @nnshah1 @rmccorm4 @GuanLuo @Tabrizian @krishung5

**Reply by nnshah1** [reply]
I'm good with a flag - I believe we initially followed the vllm example: 

https://github.com/vllm-project/vllm/blob/ed70c70ea3569670499717f06d117ed25ec32af4/vllm/entrypoints/api_server.py#L51

But we're free to change the default - 

What does the trt-llm backend do by default? (we should keep the default behavior the same between them)

**Reply by rmccorm4** [reply]
A similar consideration to make is if we should always produce the current response +1 for each response in streaming mode:
- AI
- AI is
- AI is a
- AI is a Machine 
...

Vs only producing the diffs/new tokens
- AI
- is
- a
- Machine
...

**Reply by mkhludnev** [reply]
oh. what a surprise! here is also a question https://github.com/vllm-project/vllm/issues/2659 
Here's the explanation vllm `api_server.py` just a sample  https://github.com/vllm-project/vllm/pull/1232#issuecomment-1751930329 but vLLM OpenAPI has `echo` option!
more questions in vLLM community https://github.com/vllm-project/vllm/issues/1043, https://github.com/vllm-project/vllm/issues/1612

**Reply by Tabrizian** [reply]
> What does the trt-llm backend do by default? (we should keep the default behavior the same between them)

Looks like the default behaviour for TRT-LLM is to include the prompt.

+1 on only including the diff in streaming mode.

**Reply by oandreeva-nv** [reply]
@mkhludnev trt-llm backend has `exclude_input_in_output` flag, that controls `prompt` concatenation. We will add same flag to vLLM soon

**Reply by mkhludnev** [reply]
Wow! Great. May I kick tiers under #6866 and come up with PR?

**Reply by oandreeva-nv** [reply]
@mkhludnev I am working on this PR for vLLM backend, it will be up by the end of this week. I'll tag you for review, if you are interested.
May I kindly ask you to be consistent with opening and closing issues, please. General suggestion is to keep them open until the issue is resolved. This helps with the internal task management/prioritization as well.

**Reply by mkhludnev** [reply, marked as answer]
for the reference https://github.com/triton-inference-server/vllm_backend/pull/35
